File: pbgestfinbo/pbgestfinbo-jar/src/main/java/it/csi/pbandi/pbgestfinbo/business/service/SaldoStralcioService.java
/*******************************************************************************
* Copyright Regione Piemonte - 2024
* SPDX-License-Identifier: EUPL-1.2
******************************************************************************/

package it.csi.pbandi.pbgestfinbo.business.service;

import java.security.InvalidParameterException;

import javax.servlet.http.HttpServletRequest;
import javax.ws.rs.Consumes;
import javax.ws.rs.DefaultValue;
import javax.ws.rs.GET;
import javax.ws.rs.POST;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.QueryParam;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

import it.csi.pbandi.pbgestfinbo.exception.DaoException;
import it.csi.pbandi.pbgestfinbo.integration.vo.SaldoStralcioVO;

@Path("/saldoStralcio")
public interface SaldoStralcioService {
	
	@GET
	@Path("/getSaldoStralcio")
	@Produces(MediaType.APPLICATION_JSON)
	Response getSaldoStralcio(
			@DefaultValue("0") @QueryParam("idSaldoStralcio") Long idSaldoStralcio,
			@DefaultValue("0") @QueryParam("idModalitaAgevolazione") int idModalitaAgev,
	        @Context HttpServletRequest req) throws DaoException;
	
	// inserzione saldo e stralcio	
	@POST
	@Path("/insertSaldoStralcio")
	@Consumes(MediaType.APPLICATION_JSON)
	@Produces(MediaType.APPLICATION_JSON)
	Response insertSaldoStralcio(
			SaldoStralcioVO saldoStralcio,
			@DefaultValue("0") @QueryParam("idUtente") Long idUtente,
			@DefaultValue("0") @QueryParam("idProgetto") Long idProgetto,
			@DefaultValue("0") @QueryParam("idModalitaAgevolazione") int idModalitaAgev,
			@Context HttpServletRequest request) 
			throws InvalidParameterException, Exception;
	
	// modifcare saldo e stralcio	
	@POST
	@Path("/modificaSaldoStralcio")
	@Consumes(MediaType.APPLICATION_JSON)
	@Produces(MediaType.APPLICATION_JSON)
	Response modificaSaldoStralcio(
			SaldoStralcioVO saldoStralcio,
			@DefaultValue("0") @QueryParam("idUtente") Long idUtente,
			@DefaultValue("0") @QueryParam("idProgetto") Long idProgetto,
			@DefaultValue("0") @QueryParam("idSaldoStralcio") Long idSaldoStralcio,
			@DefaultValue("0") @QueryParam("idModalitaAgevolazione") int idModalitaAgev,
			@Context HttpServletRequest request) 
					throws InvalidParameterException, Exception;
	
	// questo storico di tutti saldi e stralci
	@GET
	@Path("/getStorico")
	@Produces(MediaType.APPLICATION_JSON)
	Response getStorico(
			@DefaultValue("0") @QueryParam("idUtente") Long idUtente, 
			@DefaultValue("0") @QueryParam("idProgetto") Long idProgetto,
			@DefaultValue("0") @QueryParam("idModalitaAgevolazione") int idModalitaAgev,
		    @Context HttpServletRequest req) throws DaoException;
		
	// questo storico dei saldi con data_fine_validita == null 
	@GET
	@Path("/getStoricoSaldoStralcio")
	@Produces(MediaType.APPLICATION_JSON)
	Response getStoricoSaldoStralcio(
			@DefaultValue("0") @QueryParam("idUtente") Long idUtente, 
			@DefaultValue("0") @QueryParam("idProgetto") Long idProgetto,
			@DefaultValue("0") @QueryParam("idModalitaAgevolazione") int idModalitaAgev,
			@Context HttpServletRequest req) throws DaoException;
	
	
	@GET
	@Path("/listaAttivitaSaldoStralcio")
	@Produces(MediaType.APPLICATION_JSON)
	Response getListaAttivitaSaldoStralcio() throws DaoException;
	
	
	@GET
	@Path("/listaAttivitaEsito")
	@Produces(MediaType.APPLICATION_JSON)
	Response getListaAttivitaEsito() throws DaoException;
	
	
	@GET
	@Path("/listaAttivitaRecupero")
	@Produces(MediaType.APPLICATION_JSON)
	Response getListaAttivitaRecupero() throws DaoException;

	
	
	

}

File: pbgestfinbowcl/src/app/gestione-crediti/services/saldo-stralcio-service.ts
/*
 * Copyright Regione Piemonte - 2024 
 * SPDX-License-Identifier: EUPL-1.2 
*/

import { HttpClient, HttpParams } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { Observable } from "rxjs";
import { ConfigService } from "src/app/core/services/config.service";
import { HandleExceptionService } from "src/app/core/services/handle-exception.service";
import { AttivitaDTO } from "../commons/dto/attivita-dto";
import { SaldoStralcioVO } from "../commons/dto/saldo-stralcio-VO";
import { StoricoSaldoStralcioDTO } from "../commons/dto/storico-saldo-stralcio-dto";

@Injectable()
export class SaldoStralcioService{
    constructor(
        private configService: ConfigService,
        private http: HttpClient,
        private handleExceptionService: HandleExceptionService
    ){}

    // lista delle attivita di stralcio e saldo 
    getListaAttivitaSaldoStralcio(): Observable <Array<AttivitaDTO>>{
        let url = this.configService.getApiURL() + '/restfacade/saldoStralcio/listaAttivitaSaldoStralcio'; 
        return this.http.get<Array<AttivitaDTO>>(url);
    }

    // lista delle attivita di esito 
    getListaAttivitaEsito(): Observable <Array<AttivitaDTO>>{
        let url = this.configService.getApiURL() + '/restfacade/saldoStralcio/listaAttivitaEsito'; 
        return this.http.get<Array<AttivitaDTO>>(url);
    }
    // lista delle attivita di recupero
    getListaAttivitaRecupero(): Observable <Array<AttivitaDTO>>{
        let url = this.configService.getApiURL() + '/restfacade/saldoStralcio/listaAttivitaRecupero'; 
        return this.http.get<Array<AttivitaDTO>>(url);
    }

     // lista degli i saldi e stralci con data di fine validita == null 
     getStoricoSaldoStralcio (idUtente: number, idProgetto: number, idModalitaAgevolazione:number): Observable <Array<StoricoSaldoStralcioDTO>>{
        let url = this.configService.getApiURL() + '/restfacade/saldoStralcio/getStoricoSaldoStralcio';
        let params = new HttpParams().set("idUtente", idUtente.toString()).set("idProgetto",idProgetto.toString())
        .set("idModalitaAgevolazione", idModalitaAgevolazione.toString());
        return this.http.get<Array<StoricoSaldoStralcioDTO>>(url, { params: params });
    }

     // lista di tutti saldi e stralci 
     getStorico(idUtente: number, idProgetto: number, idModalitaAgevolazione): Observable <Array<StoricoSaldoStralcioDTO>>{
        let url = this.configService.getApiURL() + '/restfacade/saldoStralcio/getStorico';
        let params = new HttpParams().set("idUtente", idUtente.toString()).set("idProgetto",idProgetto.toString())
        .set("idModalitaAgevolazione", idModalitaAgevolazione.toString());
        return this.http.get<Array<StoricoSaldoStralcioDTO>>(url, { params: params });
    }

    getSaldoStralcio(idSaldoStralcio: number, idModalitaAgevolazione:number): Observable <SaldoStralcioVO>  {
        let url = this.configService.getApiURL() + '/restfacade/saldoStralcio/getSaldoStralcio';
         let params = new HttpParams()
         .set("idModalitaAgevolazione", idModalitaAgevolazione.toString());
        if(idSaldoStralcio){
            params = params.set("idSaldoStralcio", idSaldoStralcio.toString());
        }     
    return this.http.get<SaldoStralcioVO>(url, { params: params });
    }


    insertSaldoStralcio( saldoStralcio: SaldoStralcioVO ,idUtente: number,idProgetto: number, idModalitaAgevolazione:number): Observable <boolean> {
      
        let url = this.configService.getApiURL() + '/restfacade/saldoStralcio/insertSaldoStralcio';
        let params = new HttpParams()
        .set("idUtente", idUtente.toString())
        .set("idProgetto", idProgetto.toString())
        .set("idModalitaAgevolazione", idModalitaAgevolazione.toString());
        return this.http.post<boolean>(url, saldoStralcio, { params: params });
    }

    modificaSaldoStralcio( saldoStralcio: SaldoStralcioVO ,idUtente: number,idProgetto: number, idSaldoStralcio: number, idModalitaAgevolazione:number): Observable <boolean> {  
        let url = this.configService.getApiURL() + '/restfacade/saldoStralcio/modificaSaldoStralcio';
        let params = new HttpParams()
        .set("idUtente", idUtente.toString())
        .set("idProgetto", idProgetto.toString())
        .set("idSaldoStralcio", idSaldoStralcio.toString())
        .set("idModalitaAgevolazione", idModalitaAgevolazione.toString());
        return this.http.post<boolean>(url, saldoStralcio, { params: params });
    }

    /* insertSaldoStralcio( saldoStralcio: SaldoStralcioVO,idProgetto,allegati,idModalitaAgevolazione): Observable <boolean> {
      
        let url = this.configService.getApiURL() + '/restfacade/saldoStralcio/insertSaldoStralcio';
        let params = new HttpParams()
        let formData = new FormData();
        for (var i = 0; i < allegati.length; i++) {
            console.log(allegati[i]);
            console.log(allegati[i].name);
            formData.append("file", allegati[i], allegati[i].name);
        }
        formData.append(
            "saldoStralcio",
            JSON.stringify(saldoStralcio)
        );
        formData.append("idModalitaAgevolazione", idModalitaAgevolazione); 
        formData.append("idProgetto", idProgetto);
        return this.http.post<any>(url, formData, { params: params });
    }

    modificaSaldoStralcio( saldoStralcio: SaldoStralcioVO,idSaldoStralcio,idProgetto,allegati,idModalitaAgevolazione): Observable <boolean> {  
        let url = this.configService.getApiURL() + '/restfacade/saldoStralcio/modificaSaldoStralcio';
        let params = new HttpParams()
        let formData = new FormData();
        for (var i = 0; i < allegati.length; i++) {
            console.log(allegati[i]);
            console.log(allegati[i].name);
            formData.append("file", allegati[i], allegati[i].name);
        }
        formData.append(
            "saldoStralcio",
            JSON.stringify(saldoStralcio)
        );
        formData.append("idSaldoStralcio", idSaldoStralcio.toString()); 
        formData.append("idModalitaAgevolazione", idModalitaAgevolazione); 
        formData.append("idProgetto", idProgetto);
        return this.http.post<any>(url, formData, { params: params });
    } */


}